velox/expression/ComplexViewTypes.h

```diff
@@ -0,0 +1,355 @@
+    const std::optional<T>& rhs) {
+  return !(lhs == rhs);
+}
+
+// Represents an array of elements with an interface similar to std::vector.
+template <typename V>
+class ArrayView {
+  using reader_t = VectorReader<V, void>;
+  using element_t = typename reader_t::exec_in_t;
+
+ public:
+  ArrayView(const reader_t* reader, vector_size_t offset, vector_size_t size)
+      : reader_(reader), offset_(offset), size_(size) {}
+
+  // The previous doLoad protocol creates a value and then assigns to it.
+  // TODO: this should deprecated once we deprecate the doLoad protocol.
+  ArrayView() : reader_(nullptr), offset_(0), size_(0) {}
+
+  using Element = OptionalVectorValueAccessor<reader_t>;
+  using Iterator = IndexBasedIterator<Element>;
+
+  Iterator begin() const {
+    return Iterator{Element{reader_, offset_}};
+  }
+
+  Iterator end() const {
+    return Iterator{Element{reader_, offset_ + size_}};
+  }
+
+  // Returns true if any of the arrayViews in the vector might have null
+  // element.
+  bool mayHaveNulls() const {
+    return reader_->mayHaveNulls();
+  }
+
+  Element operator[](vector_size_t index) const {
+    return Element{reader_, index + offset_};
+  }
+
+  Element at(vector_size_t index) const {
+    return (*this)[index];
+  }
+
+  size_t size() const {
+    return size_;
+  }
+
+ private:
+  const reader_t* reader_;
+  vector_size_t offset_;
+  vector_size_t size_;
+};
+
+// This class is used to represent map inputs in simple functions with an
+// interface similar to std::map.
+template <typename K, typename V>
+class MapView {
+ public:
+  using key_reader_t = VectorReader<K, void>;
+  using value_reader_t = VectorReader<V, void>;
+  using key_element_t = typename key_reader_t::exec_in_t;
+
+  MapView(
+      const key_reader_t* keyReader,
+      const value_reader_t* valueReader,
+      vector_size_t offset,
+      vector_size_t size)
+      : keyReader_(keyReader),
+        valueReader_(valueReader),
+        offset_(offset),
+        size_(size) {}
+
+  MapView()
+      : keyReader_(nullptr), valueReader_(nullptr), offset_(0), size_(0) {}
+
+  // This class represents a lazy access wrapper around the key.
+  struct LazyKeyAccessor {
+    LazyKeyAccessor(const key_reader_t* reader, vector_size_t index)
+        : reader_(reader), index_(index) {}
+
+    operator key_element_t() const {
+      return (*reader_)[index_];
+    }
+
+    bool operator==(const LazyKeyAccessor& other) const {
+      return key_element_t(other) == key_element_t(*this);
+    }
+
+    vector_size_t index() const {
+      return index_;
+    }
+
+    void incrementIndex() {
+      index_++;
+    }
+
+   private:
+    const key_reader_t* reader_;
+    vector_size_t index_;
+  };
+
+  class Element {
+   public:
+    Element(
+        const key_reader_t* keyReader,
+        const value_reader_t* valueReader,
+        vector_size_t index)
+        : first(keyReader, index), second(valueReader, index), index_(index) {}
+    LazyKeyAccessor first;
```

Comment:
I'm somewhat concerned the indexes could get out of sync for first, second, and the enclosing Element.